### PR TITLE
Add to_char scalar function

### PR DESCRIPTION
**Pull request by autophagy**
## Summary of the changes / Why this improves CrateDB

This adds the ``to_char`` data formatting function (addresses part of #9663). This function takes an expression of either a timestamp, interval or numeric type and a string that describes how that expression should be formatted.

- Timestamps are formatted according to the syntax described in Java's DateTimeFormatter (https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html)
- Numeric values are formatted according to the syntax described in Java's DecimalFormat (https://docs.oracle.com/javase/8/docs/api/java/text/DecimalFormat.html)
- Interval values are just formatted according to the Joda default interval formatter, as the interval formatting builder has no string tokens -> formatter builder (we could, though, roll our own according to the syntax of DateTimeFormatter).

Also: should this be commited straight to master, or would a feature branch be more desirable?

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: docs/general/builtins/scalar.rst [hunk 1]
```diff
@@ -1086,6 +1086,181 @@ It has two variants depending on the type of ``timestamp``:
     +--------------+------------------+
     SELECT 1 row in set (... sec)
 
+
+``to_char(expression, format_string)``
+--------------------------------------
+
+The ``to_char`` function converts a timestamp, interval or numeric value to
```
Comment (mechanomi) [inline docs/general/builtins/scalar.rst:1093]:
```suggestion
The ``to_char`` function converts a timestamp, interval, or numeric value to
```

File: docs/general/builtins/scalar.rst [hunk 2]
```diff
@@ -1086,6 +1086,181 @@ It has two variants depending on the type of ``timestamp``:
...
+
+Synopsis
+........
+
+::
+
+    TO_CHAR( expression, format_string )
+
+``expression`` can be any value with the type of ``timestamp`` (with or without
+a timezone), ``interval`` or a ``numeric`` type.
```
Comment (mechanomi) [inline docs/general/builtins/scalar.rst:1106]:
```suggestion
a timezone), ``interval``, or a ``numeric`` type.
```

File: docs/general/builtins/scalar.rst [hunk 3]
```diff
@@ -1086,6 +1086,181 @@ It has two variants depending on the type of ``timestamp``:
...
+......
+
+The syntax for the ``format_string`` differs based the type of the expression.
+For ``timestamp`` expressions, the syntax follows that of the `Java DateTimeFormatter`_.
+For reference, the table of permitted tokens is replicated below:
+
++-------+------------------------------+------------------------------------------------+
+| Token | Meaning                      | Examples                                       |
++=======+==============================+================================================+
+| G     | Era                          | AD; Anno Domini; A                             |
```
Comment (mechanomi) [inline docs/general/builtins/scalar.rst:1118]: it might be a good idea to mark up all the tokens as literals here?

File: docs/general/builtins/scalar.rst [hunk 4]
```diff
@@ -1086,6 +1086,181 @@ It has two variants depending on the type of ``timestamp``:
...
+| '     | Escape for Text              |                                                |
++-------+------------------------------+------------------------------------------------+
+| ''    | Single Quote                 | '                                              |
++-------+------------------------------+------------------------------------------------+
+| [     | Begin Optional Section       |                                                |
++-------+------------------------------+------------------------------------------------+
+| ]     | End Optional Section         |                                                |
++-------+------------------------------+------------------------------------------------+
+
+For documentation on how to achieve the short, full and narrow forms of the
```
Comment (mechanomi) [inline docs/general/builtins/scalar.rst:1193]:
```suggestion
For documentation on how to achieve the short, full, and narrow forms of the
```

File: docs/general/builtins/scalar.rst [hunk 5]
```diff
@@ -1086,6 +1086,181 @@ It has two variants depending on the type of ``timestamp``:
...
+    +------------------------------------------------------------+
+    SELECT 1 row in set (... sec)
+
+For ``numeric`` expressions, the syntax follows that of the `Java DecimalFormat`_.
+For reference, the table of permitted tokens is replicated below:
+
++--------+---------------------+----------------------------------------------------------------------+
+| Symbol | Location            | Meaning                                                              |
++========+=====================+======================================================================+
+| 0      | Number              | Digit                                                                |
```
Comment (mechanomi) [inline docs/general/builtins/scalar.rst:1225]: same comment as above re marking up symbols as literals

**Review by mechanomi — COMMENTED** [review]
lgtm. left a few minor suggestions and comments

File: server/src/main/java/io/crate/expression/scalar/ToCharFunction.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+import io.crate.data.Input;
+import io.crate.expression.symbol.Literal;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.DataTypes;
+import io.crate.types.IntervalType;
+import org.joda.time.Period;
```
Comment (kovrus) [inline server/src/main/java/io/crate/expression/scalar/ToCharFunction.java:32]: is it absolutely necessary to use joda time? I think we were actually moving away from it in favour of java time.

File: server/src/main/java/io/crate/expression/scalar/ToCharFunction.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    @Override
+    public String evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<Object>... args) {
+        Literal expression = (Literal) args[0];
+        int dataType = expression.valueType().id();
+        if (TimestampType.ID_WITH_TZ == dataType) {
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern(String.valueOf(args[1].value()),
+                                                                      Locale.ENGLISH);
+            Long ts = DataTypes.TIMESTAMPZ.sanitizeValue(args[0].value());
```
Comment (kovrus) [inline server/src/main/java/io/crate/expression/scalar/ToCharFunction.java:82]: What do you think of using `BinaryScalar` and implementing keeping the evaluate logic a bit more separated? For example, see `CoordinateFunction` or `RoundFunction`.

File: server/src/main/java/io/crate/expression/scalar/ToCharFunction.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+    public ToCharFunction(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    @Override
+    public String evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<Object>... args) {
+        Literal expression = (Literal) args[0];
+        int dataType = expression.valueType().id();
```
Comment (kovrus) [inline server/src/main/java/io/crate/expression/scalar/ToCharFunction.java:78]: it can result in NPE. You would have to do the null check yourself or if you use `BinaryScalar`, this case will be already taken care of.

File: server/src/main/java/io/crate/expression/scalar/ToCharFunction.java (hunk 7 above)
Comment (autophagy) [reply server/src/main/java/io/crate/expression/scalar/ToCharFunction.java:82]: Aha! Yeah, I was just about to push a fixup where the evaluation logic was part of the registration step, rather than the evaluation step for every row. Thanks for the heads up :slightly_smiling_face:

File: server/src/main/java/io/crate/expression/scalar/ToCharFunction.java (hunk 6 above)
Comment (autophagy) [reply server/src/main/java/io/crate/expression/scalar/ToCharFunction.java:32]: I wasn't sure - I was following what we've used in other scalar functions that involve intervals such as `IntervalArithmeticScalar` and `IntervalTimestampArithmeticScalar`. Totally happy to change it though

File: server/src/test/java/io/crate/expression/scalar/ToCharFunctionTest.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.expression.scalar;
+
+import org.junit.Test;
+
+
+public class ToCharFunctionTest extends AbstractScalarFunctionsTest {
```
Comment (kovrus) [inline server/src/test/java/io/crate/expression/scalar/ToCharFunctionTest.java:27]: maybe we can add some tests for the corner cases, like `null` handling, etc.

File: server/src/main/java/io/crate/expression/scalar/ToCharFunction.java (hunk 6 above)
Comment (kovrus) [reply server/src/main/java/io/crate/expression/scalar/ToCharFunction.java:31]: Yes, it was used there because it seems that it was hard to go with java time in that case, but in I think the general rule we follow is going with java time if it is possible.

Comment (autophagy) [reply server/src/main/java/io/crate/expression/scalar/ToCharFunction.java:31]: Just looking into it a bit further - would it be okay to use Joda for the time being here? `IntervalType.sanitizeValue()` itself returns a Joda period, and as far as I can tell converting between a Joda Period and a Java Period isn't super straightforward. Given we also use the Joda PeriodFormat here, maybe it makes sense to stick with Joda in this instance?

File: server/src/main/java/io/crate/expression/scalar/ToCharFunction.java [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+        return formatter.format(dateTime);
+    }
+
+    private static String evaluateInterval(Object interval, String pattern) {
+        PeriodFormatter formatter = PeriodFormat.getDefault();
+        return formatter.print((Period) interval);
+    }
+
+    private static String evaluateNumber(Object number, String pattern) {
+        DecimalFormat formatter = (DecimalFormat) NumberFormat.getNumberInstance(Locale.ENGLISH);
```
Comment (kovrus) [inline server/src/main/java/io/crate/expression/scalar/ToCharFunction.java:132]: can we move `DecimalFormat` and `PeriodFormatter` out of the method scope? we do not need to init them on each evaluation, right?

File: server/src/main/java/io/crate/expression/scalar/ToCharFunction.java [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+    private final Signature signature;
+    private final Signature boundSignature;
+    private final DataType expressionType;
+    private final BiFunction<Object, String, String> evaluatorFunc;
+
+    public ToCharFunction(Signature signature, Signature boundSignature, DataType expressionType, BiFunction<Object, String, String> evaluatorFunc) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+        this.expressionType = expressionType;
```
Comment (kovrus) [inline server/src/main/java/io/crate/expression/scalar/ToCharFunction.java:113]: you do not really have to pass the argument type explicitly, you can already get it from the bound signature.
```
assert boundSignature.getArgumentDataTypes().size() == 1
boundSignature.getArgumentDataTypes().get(0)
```

File: server/src/main/java/io/crate/expression/scalar/ToCharFunction.java [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern(pattern,
+                                                                  Locale.ENGLISH);
+        Long ts = DataTypes.TIMESTAMPZ.sanitizeValue(timestamp);
+        LocalDateTime dateTime = LocalDateTime.ofInstant(Instant.ofEpochMilli(ts),
+                                                         TimeZone.getTimeZone("UTC").toZoneId());
+        return formatter.format(dateTime);
+    }
+
+    private static String evaluateInterval(Object interval, String pattern) {
+        PeriodFormatter formatter = PeriodFormat.getDefault();
```
Comment (kovrus) [inline server/src/main/java/io/crate/expression/scalar/ToCharFunction.java:127]: maybe we can reuse an instance of `PeriodFormatter` instead of initializing it on each function call.

Comment (autophagy) [reply server/src/main/java/io/crate/expression/scalar/ToCharFunction.java:127]: That would prevent the function from being ``static``, correct?

Comment (kovrus) [reply server/src/main/java/io/crate/expression/scalar/ToCharFunction.java:129]: not really, you would have to do smth like:  
```
private static final PeriodFormatter PERIOD_FORMATTER = PeriodFormat.getDefault();

```

Comment (autophagy) [reply server/src/main/java/io/crate/expression/scalar/ToCharFunction.java:129]: Ah, of course :woman_facepalming:

File: server/src/main/java/io/crate/expression/scalar/ToCharFunction.java [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+    private final DataType expressionType;
+    private final BiFunction<Object, String, String> evaluatorFunc;
+
+    private static final PeriodFormatter PERIOD_FORMATTER = PeriodFormat.getDefault();
+
+    public ToCharFunction(Signature signature, Signature boundSignature, BiFunction<Object, String, String> evaluatorFunc) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+
+        assert boundSignature.getArgumentDataTypes().size() == 2;
```
Comment (kovrus) [inline server/src/main/java/io/crate/expression/scalar/ToCharFunction.java:113]: maybe add a message to the assertion, otherwise ack

**Review by kovrus — APPROVED** [review]

**Comment by autophagy** [comment]
@kovrus Thank you for the reviews and guidance!
